Zend/zend_gc.c

```diff
@@ -1153,12 +1157,20 @@ static void gc_mark_grey(zend_refcounted *ref, gc_stack *stack)
 				}
 				zv++;
 			}
+			if (primitive_ht) {
+				GC_ADD_FLAGS(primitive_ht, GC_NOT_COLLECTABLE);
+				primitive_ht = NULL;
+			}
 		}
 	} else if (GC_TYPE(ref) == IS_ARRAY) {
 		ZEND_ASSERT(((zend_array*)ref) != &EG(symbol_table));
 		ht = (zend_array*)ref;
 handle_ht:
 		n = ht->nNumUsed;
+		if (GC_FLAGS(ht) & GC_NOT_COLLECTABLE) {
+			goto next;
+		}
+		primitive_ht = ht;
```

Comment:
This should be moved above `handle_ht`, otherwise this logic is applied to hash tables returned by `get_gc` handlers. I think it wouldn't be safe, as the ht may have INDIRECT elements, whose content can be changed without touching the array. For objects with custom `get_gc` handlers, the `ht` may be updated in unexpected ways too.

``` php
#[AllowDynamicProperties]
class C {
    public $c;
}

$c = new C();
$c->c = 1;
$c->d = 2; // dynamic prop

$x = $c;
unset($x); // add to roots

gc_collect_cycles(); // marks C->properties as not collectable

$c->c = $c; // should be collectable now, but isn't

unset($c);

gc_collect_cycles(); // ignores C->properties
```